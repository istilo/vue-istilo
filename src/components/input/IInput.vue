<script setup lang="ts">
import { computed, useAttrs, ref } from 'vue';

const input = ref();

const attrs = useAttrs();
const emit = defineEmits(['input', 'focus', 'update:modelValue']);

const props = defineProps<{
	/**
	 * Display a message and turn the form into danger color
	 */
	error?: string;

	/**
	 * Specify the fix width size
	 *
	 * @values xsmall, small, medium, large
	 */
	fixedWidth?: string;

	/**
	 * Make the border color into success color
	 */
	isSuccess?: boolean;

	/**
	 * Add a label before the input
	 */
	label?: string;

	/**
	 * The value of the component
	 */
	modelValue?: string;

	/**
	 * Add prefix custom text or icon
	 */
	prefix?: string;

	/**
	 * Increase/decrease the dimension and gutter of the component
	 */
	size?: string;

	/**
	 * Specify the width based on width component
	 *
	 * @values 1-*, 2-*, 3-^, 4-^, 5-^
	 */
	width?: string;
}>();

const labelClasses = computed(() => {
	return [
		'i-form-label',
		props.error ? 'i-text-danger' : '',
		props.isSuccess ? 'i-text-success' : '',
	];
});

const inputClasses = computed(() => {
	return [
		'i-input',
		props.isSuccess ? 'i-form-success' : '',
		props.size ? `i-form-${props.size}` : '',
		props.fixedWidth
			? `i-form-width-${props.fixedWidth}`
			: props.width
			  ? `i-width-${props.width}`
			  : '',
		props.error ? 'i-form-danger' : '',
	];
});

const handleInput = (event: any) => {
	emit('update:modelValue', event.target.value);
};
</script>

<template>
	<div class="i-form-wrapper">
		<label
			v-if="props.label"
			:class="labelClasses">
			{{ props.label
			}}<span
				v-if="attrs.required"
				:for="attrs.id"
				class="i-text-danger"
				>*
			</span>
		</label>

		<div v-if="props.prefix">
			<div class="i-flex">
				<div class="i-form-input-prefix">
					{{ props.prefix }}
				</div>

				<input
					:ref="input"
					:class="inputClasses"
					@input="handleInput"
					:type="attrs.type ? attrs.type : 'text'"
					:value="modelValue"
					:required="attrs.required"
					:id="attrs.id"
					:aria-required="attrs.required"
					:aria-placeholder="attrs.placeholder"
					:aria-label="props.label"
					:aria-labellby="props.label !== null && props.label?.length ? attrs.id : false"
					:aria-invalid="props.error ? true : null"
					role="input"
					v-bind="attrs" />
			</div>
		</div>

		<input
			v-else
			:ref="input"
			:class="inputClasses"
			@input="handleInput"
			:type="attrs.type ? attrs.type : 'text'"
			:value="modelValue"
			:required="attrs.required"
			:id="attrs.id"
			:aria-required="attrs.required"
			:aria-placeholder="attrs.placeholder"
			:aria-label="props.label"
			:aria-labellby="props.label !== null && props.label?.length ? attrs.id : false"
			:aria-invalid="props.error ? true : null"
			role="input"
			v-bind="attrs" />

		<small
			v-if="props.error"
			class="i-display-block i-margin-xsmall i-text-danger">
			{{ props.error }}
		</small>
	</div>
</template>
