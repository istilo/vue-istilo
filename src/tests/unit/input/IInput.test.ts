import { describe, it, expect, beforeEach, afterEach } from 'vitest';
import { shallowMount, VueWrapper, enableAutoUnmount } from '@vue/test-utils';

import IInput from '@/components/input/IInput.vue';

enableAutoUnmount(afterEach);

describe('Istilo Input', () => {
	let wrapper: VueWrapper<any>;

	beforeEach(() => {
		wrapper = shallowMount(IInput, {
			propsData: {
				label: 'Input',
			},
		});

		expect(wrapper.exists()).toBe(true);
	});

	afterEach(() => {
		expect(wrapper.html()).toMatchSnapshot();
	});

	it('Should render the default childs, elements and class', () => {
		expect(wrapper.classes()).toContain('i-form-wrapper');

		expect(wrapper.find('.i-form-label')).toBeDefined();
		expect(wrapper.find('.i-form-label').text()).toBe('Input');

		expect(wrapper.find('.i-input').attributes('role')).toBe('input');
		expect(wrapper.find('.i-input').attributes('aria-label')).toBe('Input');
	});

	it('Should add error modifier class and element', async () => {
		await wrapper.setProps({
			error: 'Error message',
		});

		expect(wrapper.find('.i-form-label').classes()).toContain('i-text-danger');
		expect(wrapper.find('.i-input').classes()).toContain('i-form-danger');
		expect(wrapper.find('.i-input').attributes('aria-invalid')).toBe('true');

		expect(wrapper.find('small')).toBeDefined();
		expect(wrapper.find('small').text()).toBe('Error message');
	});

	it('Should add required element and attributes', async () => {
		await wrapper.setProps({
			required: true,
		});

		expect(wrapper.find('span').text()).toBe('*');
		expect(wrapper.find('.i-input').attributes('required')).toBe('');
		expect(wrapper.find('.i-input').attributes('aria-required')).toBe('true');
	});

	it('Should add fixed width modifier class', async () => {
		await wrapper.setProps({
			fixedWidth: 'large',
		});

		expect(wrapper.find('.i-input').classes()).toContain('i-form-width-large');
	});

	it('Should add size modifier class', async () => {
		await wrapper.setProps({
			size: 'large',
		});

		expect(wrapper.find('.i-input').classes()).toContain('i-form-large');
	});

	it('Should add success modifier class', async () => {
		await wrapper.setProps({
			isSuccess: true,
		});

		expect(wrapper.find('.i-form-label').classes()).toContain('i-text-success');
		expect(wrapper.find('.i-input').classes()).toContain('i-form-success');
	});
});
