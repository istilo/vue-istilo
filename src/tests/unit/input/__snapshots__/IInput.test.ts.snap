// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Istilo Input > Should add error modifier class and element 1`] = `
"<div class=\\"i-form-wrapper\\"><label class=\\"i-form-label i-text-danger\\">Input
    <!--v-if-->
  </label><input class=\\"i-input i-form-danger\\" type=\\"text\\" aria-label=\\"Input\\" role=\\"input\\" aria-invalid=\\"true\\"><small class=\\"i-display-block i-margin-xsmall i-text-danger\\">Error message</small></div>"
`;

exports[`Istilo Input > Should add fixed width modifier class 1`] = `
"<div class=\\"i-form-wrapper\\"><label class=\\"i-form-label\\">Input
    <!--v-if-->
  </label><input class=\\"i-input i-form-width-large\\" type=\\"text\\" aria-label=\\"Input\\" role=\\"input\\">
  <!--v-if-->
</div>"
`;

exports[`Istilo Input > Should add required element and attributes 1`] = `
"<div class=\\"i-form-wrapper\\" required=\\"true\\"><label class=\\"i-form-label\\">Input<span class=\\"i-text-danger\\">* </span></label><input class=\\"i-input\\" type=\\"text\\" aria-label=\\"Input\\" role=\\"input\\" required=\\"\\" aria-required=\\"true\\">
  <!--v-if-->
</div>"
`;

exports[`Istilo Input > Should add size modifier class 1`] = `
"<div class=\\"i-form-wrapper\\"><label class=\\"i-form-label\\">Input
    <!--v-if-->
  </label><input class=\\"i-input i-form-large\\" type=\\"text\\" aria-label=\\"Input\\" role=\\"input\\">
  <!--v-if-->
</div>"
`;

exports[`Istilo Input > Should add success modifier class 1`] = `
"<div class=\\"i-form-wrapper\\"><label class=\\"i-form-label i-text-success\\">Input
    <!--v-if-->
  </label><input class=\\"i-input i-form-success\\" type=\\"text\\" aria-label=\\"Input\\" role=\\"input\\">
  <!--v-if-->
</div>"
`;

exports[`Istilo Input > Should render the default childs, elements and class 1`] = `
"<div class=\\"i-form-wrapper\\"><label class=\\"i-form-label\\">Input
    <!--v-if-->
  </label><input class=\\"i-input\\" type=\\"text\\" aria-label=\\"Input\\" role=\\"input\\">
  <!--v-if-->
</div>"
`;
